File: backend/src/services/userService.ts
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { JwtPayload } from '../types';
import { createError } from '../middleware/errorHandler';

const prisma = new PrismaClient();

export async function registerUser(email: string, password: string, name: string, phone?: string) {
  const existing = await prisma.user.findUnique({ where: { email } });
  if (existing) throw createError('Email already registered', 409);

  if (password.length < 6) throw createError('Password must be at least 6 characters', 400);

  const passwordHash = await bcrypt.hash(password, 10);
  const user = await prisma.user.create({
    data: { email, passwordHash, name, phone },
    select: { id: true, email: true, name: true, phone: true, role: true, createdAt: true },
  });

  await prisma.botConfig.create({
    data: {
      userId: user.id,
      confidenceThreshold: parseFloat(process.env.DEFAULT_CONFIDENCE_THRESHOLD ?? '0.65'),
      maxOrderSizeTry: parseFloat(process.env.DEFAULT_MAX_ORDER_SIZE_TRY ?? '10000'),
      dailyLossLimitTry: parseFloat(process.env.DEFAULT_DAILY_LOSS_LIMIT_TRY ?? '5000'),
      tickIntervalMs: parseInt(process.env.BOT_TICK_INTERVAL_MS ?? '5000'),
      watchlist: ['THYAO', 'GARAN', 'AKBNK'],
    },
  });

  return { user, tokens: generateTokens(user) };
}

export async function loginUser(email: string, password: string) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw createError('Invalid email or password', 401);

  const valid = await bcrypt.compare(password, user.passwordHash);
  if (!valid) throw createError('Invalid email or password', 401);

  const { passwordHash: _, ...safeUser } = user;
  return { user: safeUser, tokens: generateTokens(safeUser) };
}

export async function getUserById(userId: string) {
  const user = await prisma.user.findUnique({
    where: { id: userId },
    select: { id: true, email: true, name: true, phone: true, role: true, createdAt: true },
  });
  if (!user) throw createError('User not found', 404);
  return user;
}

function generateTokens(user: { id: string; email: string; role: string }) {
  const secret = process.env.JWT_SECRET ?? 'fallback-secret';
  const payload: JwtPayload = { userId: user.id, email: user.email, role: user.role };

  const accessToken = jwt.sign(payload, secret, { expiresIn: '7d' });
  const refreshToken = jwt.sign(payload, secret, { expiresIn: '30d' });

  return { accessToken, refreshToken };
}

export async function refreshTokens(refreshToken: string) {
  const secret = process.env.JWT_SECRET ?? 'fallback-secret';
  try {
    const payload = jwt.verify(refreshToken, secret) as JwtPayload;
    const user = await getUserById(payload.userId);
    return generateTokens(user);
  } catch (err) {
    console.error('[UserService] Refresh token invalid:', err);
    throw createError('Invalid refresh token', 401);
  }
}

File: backend/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using TradingBot.Data;
using TradingBot.DTOs;
using TradingBot.Models;

namespace TradingBot.Services;

public class UserService(AppDbContext db, JwtService jwt, IConfiguration config)
{
    public async Task<AuthResponse> RegisterAsync(RegisterRequest req)
    {
        if (await db.Users.AnyAsync(u => u.Email == req.Email))
            throw new AppException("Email already registered", 409);

        if (req.Password.Length < 6)
            throw new AppException("Password must be at least 6 characters", 400);

        var user = new User
        {
            Email = req.Email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(req.Password),
            Name = req.Name,
            Phone = req.Phone,
        };
        db.Users.Add(user);

        db.BotConfigs.Add(new BotConfig
        {
            UserId = user.Id,
            ConfidenceThreshold = config.GetValue<double>("Trading:ConfidenceThreshold", 0.65),
            MaxOrderSizeTry = config.GetValue<double>("Trading:MaxOrderSizeTry", 10000),
            DailyLossLimitTry = config.GetValue<double>("Trading:DailyLossLimitTry", 5000),
            TickIntervalMs = config.GetValue<int>("Trading:TickIntervalMs", 5000),
            Watchlist = ["THYAO", "GARAN", "AKBNK"],
        });

        await db.SaveChangesAsync();
        var (access, refresh) = jwt.GenerateTokens(user);
        return new AuthResponse(ToDto(user), new TokensDto(access, refresh));
    }

    public async Task<AuthResponse> LoginAsync(LoginRequest req)
    {
        var user = await db.Users.FirstOrDefaultAsync(u => u.Email == req.Email);
        if (user == null || !BCrypt.Net.BCrypt.Verify(req.Password, user.PasswordHash))
            throw new AppException("Invalid email or password", 401);

        var (access, refresh) = jwt.GenerateTokens(user);
        return new AuthResponse(ToDto(user), new TokensDto(access, refresh));
    }

    public async Task<UserDto> GetByIdAsync(Guid userId)
    {
        var user = await db.Users.FindAsync(userId)
            ?? throw new AppException("User not found", 404);
        return ToDto(user);
    }

    public async Task<TokensDto> RefreshAsync(string refreshToken)
    {
        var principal = jwt.ValidateToken(refreshToken)
            ?? throw new AppException("Invalid refresh token", 401);

        var userIdStr = principal.FindFirst("userId")?.Value
            ?? throw new AppException("Invalid token claims", 401);

        var user = await db.Users.FindAsync(Guid.Parse(userIdStr))
            ?? throw new AppException("User not found", 404);

        var (access, refresh) = jwt.GenerateTokens(user);
        return new TokensDto(access, refresh);
    }

    private static UserDto ToDto(User u) => new(u.Id, u.Email, u.Name, u.Phone, u.Role, u.CreatedAt);
}

public class AppException(string message, int statusCode = 400) : Exception(message)
{
    public int StatusCode { get; } = statusCode;
}
